pkg/function/service.go

```diff
@@ -132,6 +134,44 @@ func (fs *Service) MapFn(ctx context.Context, d *functionpb.DatumRequest) (*func
 	return datumList, nil
 }
 
+// MapStreamFn applies a function to each datum element and returns a stream.
+func (fs *Service) MapStreamFn(d *functionpb.DatumRequest, stream functionpb.UserDefinedFunction_MapStreamFnServer) error {
+	var hd = handlerDatum{
+		value:     d.GetValue(),
+		eventTime: d.GetEventTime().EventTime.AsTime(),
+		watermark: d.GetWatermark().Watermark.AsTime(),
+		metadata: handlerDatumMetadata{
+			id:           d.GetMetadata().GetId(),
+			numDelivered: d.GetMetadata().GetNumDelivered(),
+		},
+	}
+	ctx := stream.Context()
+	messageCh := make(chan Message)
+
+	go func() {
+		fs.MapperStream.HandleDo(ctx, d.GetKeys(), &hd, messageCh)
+		// Close the channel when not closed by the user.
+		if !isClosed(messageCh) {
```

Comment:
To my understanding, `isClosed` is very tricky in Golang. An alternative I can think of is, using `for/select` to read messages, and check if the goroutine above is done. Think it through to see what is the best way.

Reply:
Yeah, with that we will leave the channel open which probably would be fine as the channel would be GCed, and the data referenced by the channel should also be GCed. Although I would prefer not to do so as it is not a good practice, and there is an alternative to properly close the channel.

Reply:
I agree, the channel can be GCed, we should focus on not losing any messages. Will following code work?

```golang
	done := make(chan bool)
	go func() {
		fs.MapperStream.HandleDo(ctx, d.GetKeys(), &hd, messageCh)
		done <- true
	}
	finished := false
	for {
		select {
			case <-closeCh:
				finished = true
			case message, ok := <- range messageCh:
				if !ok {
					close(done)
					return nil
				}
				// send message
			default:
				if finished {
					close(messageCh)
					close(done)
					return nil
				}
		}
	}
```

Reply:
the problem of this is 1) we cannot close the channel in the receiver go routine as `isClosed` won't work as we discussed. 2) without checking if the channel is closed, we cannot close the channel safely. So we have to leave it to GC.

Reply:
> the problem of this is 1) we cannot close the channel in the receiver go routine as `isClosed` won't work as we discussed. 2) without checking if the channel is closed, we cannot close the channel safely. So we have to leave it to GC.

With the code above, we do not need to check if the channel is closed.

2 use cases:

1. Channel is closed by the UDF, then it will end in the branch `case message, ok := <- messageCh:` after reading all the messages;
2. Channel is not closed by the UDF, then it will go to `case <-done` first, and it will go to `default`, and it's safe to `close(messageCh)`.

Reply:
Check the sample code https://go.dev/play/p/JOk8nxyDXYw.

Reply:
Sounds good! As we discussed offline, even though both approach work, I'm updating it per your suggestions to make it easier to follow. Thanks!